File: services/voice-ai/extract.py
"""
Structured extraction from a live call transcript, via local Ollama.

The prompt is deliberately hostile toward guessing. A wrong auto-written email
costs a deal, so every non-null field must carry the prospect's literal words
as evidence — if the model cannot quote it, the field is null by construction.
"""

from __future__ import annotations

import json
import os
import re
from datetime import datetime, timedelta
from typing import Any
from zoneinfo import ZoneInfo

import httpx

OLLAMA_URL = os.environ.get("OLLAMA_URL", "http://localhost:11434")
MODEL = os.environ.get("OLLAMA_MODEL", "qwen2.5:7b-instruct")

SYSTEM_PROMPT = """You extract structured facts from a live B2B cold call transcript. You output JSON only. No prose, no markdown, no code fences.

CONTEXT PROVIDED TO YOU:
- current_datetime_operator_tz: the exact current date and time in the operator's timezone
- operator_timezone: IANA timezone of the operator's calendar
- prospect_area_code: the area code of the number being dialed
- prospect_inferred_timezone: timezone derived from that area code
- lead_on_file: the currently stored first_name, last_name, email, company, address
- transcript_window: the last 8 speaker-tagged turns

ABSOLUTE RULES:
1. Extract ONLY what the PROSPECT explicitly stated. Never extract from OPERATOR turns. The operator reading an email address aloud to confirm it is NOT the prospect providing it.
2. If a fact was not explicitly stated, output null. Never infer, never complete, never guess a plausible value.
3. For every non-null field you MUST populate "evidence" with the prospect's literal words that justify it. If you cannot quote them, the field must be null.
4. Output a confidence between 0.0 and 1.0 for every non-null field. Be harsh. Audio is noisy and mistakes are expensive.
5. Only report a field as changed if it MEANINGFULLY differs from lead_on_file. Ignore casing, punctuation, and whitespace differences.

EMAIL RULES:
- Reconstruct spelled-out addresses: "j-o-s-h at gmail dot com" becomes "josh@gmail.com".
- "dot" becomes ".", "at" becomes "@", "underscore" becomes "_", "dash" and "hyphen" become "-".
- Only report an email correction if the prospect is supplying a DIFFERENT address, not confirming the existing one. "Yes that's right" is a confirmation — output null.
- Validate the result is a syntactically valid address. If not, output null.

DATETIME RULES:
- Resolve relative expressions against current_datetime_operator_tz. "Next Tuesday" means the Tuesday of the following week, not the coming Tuesday, unless the prospect clarifies.
- Determine the prospect's timezone in this priority order: (a) they explicitly state one, (b) prospect_inferred_timezone from area code, (c) operator_timezone. Report which method you used in timezone_source.
- Convert the agreed time INTO operator_timezone and output it in that zone as an ISO 8601 string with offset.
- Ambiguous AM/PM: business hours assumption only for 8-11 (AM) and 1-6 (PM). For 12 or 7, output null and set needs_clarification true.
- A time is only a booking if the prospect AGREED to it. A time the operator proposed and the prospect did not accept is not a booking.

STAGE RULES — choose exactly one:
- "booked": prospect agreed to a specific date AND time
- "interested": positive signal, wants information, asked substantive questions, but no time agreed
- "callback": asked to be contacted at a later unspecified time, or said now is bad
- "not_interested": explicit rejection, asked to be removed, or hostile
- "new": no clear signal yet
Do not report "booked" without a resolvable datetime.

OUTPUT SCHEMA — return exactly this shape:
{
  "email": {"value": string|null, "evidence": string|null, "confidence": number},
  "first_name": {"value": string|null, "evidence": string|null, "confidence": number},
  "last_name": {"value": string|null, "evidence": string|null, "confidence": number},
  "company": {"value": string|null, "evidence": string|null, "confidence": number},
  "address": {"value": string|null, "evidence": string|null, "confidence": number},
  "booking": {
    "datetime_operator_tz": string|null,
    "prospect_stated_time": string|null,
    "timezone_source": "explicit"|"area_code"|"operator_default"|null,
    "duration_minutes": number|null,
    "needs_clarification": boolean,
    "evidence": string|null,
    "confidence": number
  },
  "stage": {"value": string, "evidence": string|null, "confidence": number},
  "is_gatekeeper": boolean,
  "is_voicemail_or_automated": boolean
}"""

VERIFY_PROMPT = """You are verifying a proposed appointment extraction. Given the transcript window, the current datetime in the operator's timezone, and the proposed booking, answer:
1. Did the prospect explicitly agree to this specific date and time?
2. Is the timezone conversion arithmetic correct?
3. Is the resolved date in the future and within the next 90 days?
Output JSON only: {"verified": boolean, "reason": string, "corrected_datetime": string|null}"""


# --- cheap gate -------------------------------------------------------------
# Running a 7B model on every segment wastes compute and adds latency for no
# benefit — most turns contain nothing extractable. These heuristics decide
# whether a segment is even worth asking about.

_EMAIL_HINTS = re.compile(
    r"@|at gmail|at yahoo|at outlook|at hotmail|dot com|dot net|dot org|dot co",
    re.I,
)
_SPELLED = re.compile(r"\b(?:[a-z][\s,.-]+){3,}", re.I)
_TIME_HINTS = re.compile(
    r"\b(monday|tuesday|wednesday|thursday|friday|saturday|sunday|tomorrow|"
    r"next week|morning|afternoon|evening|noon|"
    r"january|february|march|april|may|june|july|august|september|october|"
    r"november|december|"
    r"\d{1,2}\s*(?:am|pm)|\d{1,2}:\d{2})\b",
    re.I,
)
_INTENT = re.compile(
    r"\b(not interested|remove me|take me off|call me back|reach me|send me|"
    r"sounds good|let'?s do it|no thanks|stop calling)\b",
    re.I,
)
_PHONE = re.compile(r"(?:\d[\s-]?){7,}")


def worth_extracting(text: str) -> bool:
    """True when a prospect segment plausibly contains an extractable fact."""
    return bool(
        _EMAIL_HINTS.search(text)
        or _SPELLED.search(text)
        or _TIME_HINTS.search(text)
        or _INTENT.search(text)
        or _PHONE.search(text)
    )


async def _ollama(system: str, user: str) -> dict[str, Any]:
    async with httpx.AsyncClient(timeout=45.0) as client:
        res = await client.post(
            f"{OLLAMA_URL}/api/chat",
            json={
                "model": MODEL,
                "messages": [
                    {"role": "system", "content": system},
                    {"role": "user", "content": user},
                ],
                "stream": False,
                "format": "json",
                "options": {"temperature": 0},
            },
        )
        res.raise_for_status()
        content = res.json()["message"]["content"]
        return json.loads(content)


async def extract(
    transcript_window: list[dict[str, str]],
    lead_on_file: dict[str, Any],
    operator_tz: str = "America/Vancouver",
    prospect_area_code: str | None = None,
    prospect_tz: str | None = None,
) -> dict[str, Any]:
    now = datetime.now(ZoneInfo(operator_tz))

    payload = {
        "current_datetime_operator_tz": now.isoformat(),
        "operator_timezone": operator_tz,
        "prospect_area_code": prospect_area_code,
        "prospect_inferred_timezone": prospect_tz or operator_tz,
        "lead_on_file": lead_on_file,
        "transcript_window": transcript_window[-8:],
    }

    return await _ollama(SYSTEM_PROMPT, json.dumps(payload))


async def verify_booking(
    transcript_window: list[dict[str, str]],
    proposed_iso: str,
    operator_tz: str = "America/Vancouver",
) -> dict[str, Any]:
    """Second pass (§5.5). Booking errors are the expensive kind, so a proposal
    is only surfaced to the operator once a fresh call agrees with it."""
    now = datetime.now(ZoneInfo(operator_tz))

    payload = {
        "current_datetime_operator_tz": now.isoformat(),
        "operator_timezone": operator_tz,
        "proposed_booking": proposed_iso,
        "transcript_window": transcript_window[-8:],
    }

    result = await _ollama(VERIFY_PROMPT, json.dumps(payload))

    # Belt and braces: enforce the 90-day window in code rather than trusting
    # the model's arithmetic, which is the part it is worst at.
    candidate = result.get("corrected_datetime") or proposed_iso
    try:
        when = datetime.fromisoformat(candidate)
        if when <= now or when > now + timedelta(days=90):
            return {
                "verified": False,
                "reason": "Resolved date is in the past or beyond 90 days.",
                "corrected_datetime": None,
            }
    except (ValueError, TypeError):
        return {
            "verified": False,
            "reason": "Could not parse the resolved datetime.",
            "corrected_datetime": None,
        }

    return result

File: apps/web/src/integrations/ai/extract.ts
import { OPERATOR_TIMEZONE } from '@actualizecrm/db';
import { complete } from './client';
import {
  EXTRACTION_SYSTEM,
  VERIFY_SYSTEM,
  ANALYSIS_SYSTEM,
  worthExtracting,
} from './prompts';
import { areaCodeOf } from '@/lib/phone';

export interface Field<T = string> {
  value: T | null;
  evidence: string | null;
  confidence: number;
}

export interface BookingField {
  datetime_operator_tz: string | null;
  prospect_stated_time: string | null;
  timezone_source: 'explicit' | 'area_code' | 'operator_default' | null;
  duration_minutes: number | null;
  needs_clarification: boolean;
  evidence: string | null;
  confidence: number;
}

export interface Extraction {
  email: Field;
  first_name: Field;
  last_name: Field;
  company: Field;
  address: Field;
  booking: BookingField;
  stage: Field;
  is_gatekeeper: boolean;
  is_voicemail_or_automated: boolean;
}

export interface TranscriptTurn {
  speaker: 'operator' | 'prospect';
  text: string;
}

export interface LeadOnFile {
  first_name?: string | null;
  last_name?: string | null;
  email?: string | null;
  company?: string | null;
  address?: string | null;
}

const EMPTY_FIELD: Field = { value: null, evidence: null, confidence: 0 };

const EMPTY_EXTRACTION: Extraction = {
  email: EMPTY_FIELD,
  first_name: EMPTY_FIELD,
  last_name: EMPTY_FIELD,
  company: EMPTY_FIELD,
  address: EMPTY_FIELD,
  booking: {
    datetime_operator_tz: null,
    prospect_stated_time: null,
    timezone_source: null,
    duration_minutes: null,
    needs_clarification: false,
    evidence: null,
    confidence: 0,
  },
  stage: { value: 'new', evidence: null, confidence: 0 },
  is_gatekeeper: false,
  is_voicemail_or_automated: false,
};

/// Rough area-code → IANA zone map for the NANP zones an operator actually
/// dials. Anything unlisted falls back to the operator's own timezone, which
/// the prompt is told to report as `operator_default`.
const AREA_CODE_TZ: Record<string, string> = {
  // Pacific
  '206': 'America/Los_Angeles', '213': 'America/Los_Angeles', '236': 'America/Vancouver',
  '250': 'America/Vancouver', '310': 'America/Los_Angeles', '323': 'America/Los_Angeles',
  '408': 'America/Los_Angeles', '415': 'America/Los_Angeles', '503': 'America/Los_Angeles',
  '509': 'America/Los_Angeles', '604': 'America/Vancouver', '619': 'America/Los_Angeles',
  '702': 'America/Los_Angeles', '714': 'America/Los_Angeles', '778': 'America/Vancouver',
  '858': 'America/Los_Angeles', '925': 'America/Los_Angeles',
  // Mountain
  '303': 'America/Denver', '385': 'America/Denver', '480': 'America/Phoenix',
  '505': 'America/Denver', '602': 'America/Phoenix', '720': 'America/Denver',
  '801': 'America/Denver', '403': 'America/Edmonton', '587': 'America/Edmonton',
  // Central
  '214': 'America/Chicago', '281': 'America/Chicago', '312': 'America/Chicago',
  '512': 'America/Chicago', '612': 'America/Chicago', '713': 'America/Chicago',
  '773': 'America/Chicago', '832': 'America/Chicago', '204': 'America/Winnipeg',
  // Eastern
  '212': 'America/New_York', '215': 'America/New_York', '305': 'America/New_York',
  '404': 'America/New_York', '416': 'America/Toronto', '617': 'America/New_York',
  '646': 'America/New_York', '703': 'America/New_York', '718': 'America/New_York',
  '813': 'America/New_York', '905': 'America/Toronto',
};

function timezoneForAreaCode(phone: string | null): {
  areaCode: string | null;
  tz: string;
} {
  const areaCode = phone ? areaCodeOf(phone) : null;
  return {
    areaCode,
    tz: (areaCode && AREA_CODE_TZ[areaCode]) || OPERATOR_TIMEZONE,
  };
}

/**
 * Live extraction over the trailing transcript window.
 *
 * Returns null when the segment does not clear the cheap gate — the caller
 * should treat that as "nothing to show", not as a failure.
 */
export async function extractFromWindow(
  window: TranscriptTurn[],
  leadOnFile: LeadOnFile,
  prospectPhone: string | null,
  opts: { force?: boolean } = {},
): Promise<Extraction | null> {
  const lastProspect = [...window]
    .reverse()
    .find((t) => t.speaker === 'prospect');

  if (!opts.force) {
    if (!lastProspect) return null;
    if (!worthExtracting(lastProspect.text)) return null;
  }

  const { areaCode, tz } = timezoneForAreaCode(prospectPhone);

  const payload = {
    current_datetime_operator_tz: new Date().toLocaleString('sv-SE', {
      timeZone: OPERATOR_TIMEZONE,
    }),
    operator_timezone: OPERATOR_TIMEZONE,
    prospect_area_code: areaCode,
    prospect_inferred_timezone: tz,
    lead_on_file: leadOnFile,
    transcript_window: window.slice(-8),
  };

  const { value } = await complete<Extraction>(
    EXTRACTION_SYSTEM,
    JSON.stringify(payload),
    EMPTY_EXTRACTION,
  );

  return value;
}

export interface Verification {
  verified: boolean;
  reason: string;
  corrected_datetime: string | null;
}

/**
 * Second pass over a proposed booking (§5.5).
 *
 * A mis-parsed timezone puts a meeting in the wrong hour and the prospect is
 * gone, so a proposal is only surfaced once a fresh call agrees with it. The
 * date-window check is enforced here in code rather than trusted to the model,
 * because arithmetic is the part it is worst at.
 */
export async function verifyBooking(
  window: TranscriptTurn[],
  proposedIso: string,
): Promise<Verification> {
  const payload = {
    current_datetime_operator_tz: new Date().toLocaleString('sv-SE', {
      timeZone: OPERATOR_TIMEZONE,
    }),
    operator_timezone: OPERATOR_TIMEZONE,
    proposed_booking: proposedIso,
    transcript_window: window.slice(-8),
  };

  const { value } = await complete<Verification>(
    VERIFY_SYSTEM,
    JSON.stringify(payload),
    { verified: false, reason: 'Verification failed to parse.', corrected_datetime: null },
    512,
  );

  const candidate = value.corrected_datetime || proposedIso;
  const when = new Date(candidate);
  const now = Date.now();
  const ninetyDays = now + 90 * 24 * 60 * 60 * 1000;

  if (Number.isNaN(when.getTime())) {
    return {
      verified: false,
      reason: 'Could not parse the resolved datetime.',
      corrected_datetime: null,
    };
  }
  if (when.getTime() <= now || when.getTime() > ninetyDays) {
    return {
      verified: false,
      reason: 'Resolved date is in the past or beyond 90 days.',
      corrected_datetime: null,
    };
  }

  return value;
}

export interface CallAnalysis {
  summary: string;
  outcome:
    | 'booked' | 'interested' | 'callback' | 'not_interested'
    | 'no_contact' | 'gatekeeper' | 'voicemail';
  talk_ratio_operator: number;
  objections: { type: string; quote: string; handled: boolean }[];
  buying_signals: { signal: string; quote: string }[];
  coaching: { moment: string; what_happened: string; try_instead: string }[];
  next_step: string | null;
  sentiment: 'positive' | 'neutral' | 'negative';
  prospect_engaged: boolean;
}

const EMPTY_ANALYSIS: CallAnalysis = {
  summary: '',
  outcome: 'no_contact',
  talk_ratio_operator: 0,
  objections: [],
  buying_signals: [],
  coaching: [],
  next_step: null,
  sentiment: 'neutral',
  prospect_engaged: false,
};

/// End-of-call analysis over the whole transcript. Runs once per call.
export async function analyseCall(
  transcript: TranscriptTurn[],
  meta: { durationSec: number; disposition?: string | null },
): Promise<CallAnalysis | null> {
  // A transcript this short has nothing to analyse and would only produce
  // confident-sounding noise.
  if (transcript.length < 2) return null;

  const payload = {
    duration_seconds: meta.durationSec,
    operator_disposition: meta.disposition ?? null,
    transcript,
  };

  const { value } = await complete<CallAnalysis>(
    ANALYSIS_SYSTEM,
    JSON.stringify(payload),
    EMPTY_ANALYSIS,
    3072,
  );

  return value;
}
